File: trifecta-ios/Trifecta/Core/Stores/ThreadStore.swift
import Foundation
import Observation

struct ThreadDiffChange: Identifiable, Hashable {
    let id: String
    let title: String
    let command: String?
    let detail: String?
    let files: [String]
    let createdAt: Date
}

@Observable
final class ThreadStore {
    let threadId: ThreadID
    var detail: ThreadDetail?
    var messages: [Message] = []
    var session: OrchestrationSession?
    var proposedPlans: [ProposedPlan] = []
    var activities: [ThreadActivity] = []
    var pendingApprovals: [PendingApproval] = []
    var pendingUserInputs: [PendingUserInput] = []
    var lastError: String?
    var isSending: Bool = false

    private var subscription: StreamSubscription?
    private weak var client: T3Client?

    init(threadId: ThreadID) {
        self.threadId = threadId
    }

    func start(client: T3Client) async {
        self.client = client
        subscription = try? await client.subscribeThread(threadId: threadId) { item in
            Task { @MainActor [weak self] in
                self?.handle(item: item)
            }
        }
    }

    func stop() async {
        if let sub = subscription { await sub.cancel() }
        subscription = nil
    }

    var latestProposedPlan: ProposedPlan? {
        let sortedPlans = proposedPlans.sorted { $0.updatedAt < $1.updatedAt }
        guard let plan = sortedPlans.last else { return nil }
        guard plan.implementedAt == nil else { return nil }
        return plan
    }

    var isTurnRunning: Bool {
        if let session, session.status == .running { return true }
        if let state = detail?.latestTurn?.state, state == .running { return true }
        return isSending
    }

    var diffChanges: [ThreadDiffChange] {
        let rendered = RenderableActivity.collapse(activities)
        return rendered
            .filter { !$0.changedFiles.isEmpty }
            .map {
                ThreadDiffChange(
                    id: $0.id,
                    title: $0.title,
                    command: $0.command,
                    detail: $0.detail,
                    files: $0.changedFiles,
                    createdAt: $0.createdAt
                )
            }
            .sorted { $0.createdAt > $1.createdAt }
    }

    func sendMessage(text: String,
                     attachments: [UploadImage],
                     fallbackModelSelection: ModelSelection?) async {
        guard let client else { return }
        guard !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
                || !attachments.isEmpty else { return }

        let resolvedSelection = detail?.modelSelection ?? fallbackModelSelection
        guard let modelSelection = resolvedSelection else {
            await MainActor.run { self.lastError = "No model selected" }
            return
        }

        await MainActor.run { self.isSending = true }
        defer { Task { @MainActor in self.isSending = false } }

        do {
            try await client.dispatchTurnStart(
                threadId: threadId,
                text: text,
                attachments: attachments,
                modelSelection: modelSelection,
                runtimeMode: detail?.runtimeMode ?? .fullAccess,
                interactionMode: detail?.interactionMode ?? .default
            )
        } catch {
            await MainActor.run { self.lastError = error.localizedDescription }
        }
    }

    func interruptTurn() async {
        guard let client else { return }
        do {
            try await client.interruptTurn(threadId: threadId,
                                           turnId: detail?.latestTurn?.turnId)
        } catch {
            await MainActor.run { self.lastError = error.localizedDescription }
        }
    }

    func respondApproval(_ approval: PendingApproval, decision: ApprovalDecision) async {
        guard let client else { return }
        await MainActor.run {
            self.pendingApprovals.removeAll { $0.requestId == approval.requestId }
        }
        do {
            try await client.respondApproval(threadId: threadId,
                                             requestId: approval.requestId,
                                             decision: decision)
        } catch {
            await MainActor.run { self.lastError = error.localizedDescription }
        }
    }

    func respondUserInput(_ input: PendingUserInput,
                          answers: [String: Any]) async {
        guard let client else { return }
        await MainActor.run {
            self.pendingUserInputs.removeAll { $0.requestId == input.requestId }
        }
        do {
            try await client.respondUserInput(threadId: threadId,
                                              requestId: input.requestId,
                                              answers: answers)
        } catch {
            await MainActor.run { self.lastError = error.localizedDescription }
        }
    }

    func setInteractionMode(_ mode: ProviderInteractionMode) async {
        guard let client, detail?.interactionMode != mode else { return }
        do {
            try await client.setInteractionMode(threadId: threadId, mode: mode)
        } catch {
            await MainActor.run { self.lastError = error.localizedDescription }
        }
    }

    func setRuntimeMode(_ mode: RuntimeMode) async {
        guard let client, detail?.runtimeMode != mode else { return }
        do {
            try await client.setRuntimeMode(threadId: threadId, mode: mode)
        } catch {
            await MainActor.run { self.lastError = error.localizedDescription }
        }
    }

    func updateModelSelection(_ selection: ModelSelection) async {
        await MainActor.run {
            if var detail = self.detail {
                detail.modelSelection = selection
                self.detail = detail
            }
        }
        guard let client else { return }
        do {
            try await client.updateThreadModelSelection(threadId: threadId,
                                                        modelSelection: selection)
        } catch {
            await MainActor.run { self.lastError = error.localizedDescription }
        }
    }

    func implementProposedPlan(_ plan: ProposedPlan) async {
        guard let client, let detail else { return }
        do {
            try await client.startTurnFromProposedPlan(
                threadId: threadId,
                planId: plan.id,
                sourceThreadId: threadId,
                modelSelection: detail.modelSelection,
                runtimeMode: detail.runtimeMode,
                interactionMode: .default
            )
        } catch {
            await MainActor.run { self.lastError = error.localizedDescription }
        }
    }

    @MainActor
    private func handle(item: ThreadStreamItem) {
        switch item {
        case .snapshot(let detail, _):
            self.detail = detail
            self.messages = detail.messages.sorted { $0.createdAt < $1.createdAt }
            self.session = detail.session
            self.proposedPlans = detail.proposedPlans
            self.activities = detail.activities
            recomputePending()
        case .event(let event):
            apply(event)
        }
    }

    @MainActor
    private func apply(_ event: ThreadEvent) {
        switch event.type {
        case "thread.message-sent":
            applyMessageSent(event)
        case "thread.session-set":
            if let sessionDict = event.payload["session"] as? [String: Any],
               let data = try? JSONSerialization.data(withJSONObject: sessionDict),
               let session = try? JSONDecoder().decode(OrchestrationSession.self, from: data) {
                self.session = session
                if var detail = self.detail {
                    detail.session = session
                    self.detail = detail
                }
            }
        case "thread.meta-updated":
            applyMetaUpdated(event)
        case "thread.runtime-mode-set":
            if let raw = event.payload["runtimeMode"] as? String,
               let mode = RuntimeMode(rawValue: raw),
               var detail = self.detail {
                detail.runtimeMode = mode
                self.detail = detail
            }
        case "thread.interaction-mode-set":
            if let raw = event.payload["interactionMode"] as? String,
               let mode = ProviderInteractionMode(rawValue: raw),
               var detail = self.detail {
                detail.interactionMode = mode
                self.detail = detail
            }
        case "thread.archived":
            if var detail = self.detail {
                detail.archivedAt = (event.payload["archivedAt"] as? String).flatMap(ISO8601Decoder.parse) ?? Date()
                self.detail = detail
            }
        case "thread.unarchived":
            if var detail = self.detail {
                detail.archivedAt = nil
                self.detail = detail
            }
        case "thread.activity-appended":
            if let activityDict = event.payload["activity"] as? [String: Any],
               let activity = ThreadActivity.decode(from: activityDict) {
                upsertActivity(activity)
                recomputePending()
            }
        case "thread.proposed-plan-upserted":
            if let planDict = event.payload["proposedPlan"] as? [String: Any],
               let plan = ProposedPlan.decode(from: planDict) {
                if let i = proposedPlans.firstIndex(where: { $0.id == plan.id }) {
                    proposedPlans[i] = plan
                } else {
                    proposedPlans.append(plan)
                }
                if var detail = self.detail {
                    if let i = detail.proposedPlans.firstIndex(where: { $0.id == plan.id }) {
                        detail.proposedPlans[i] = plan
                    } else {
                        detail.proposedPlans.append(plan)
                    }
                    self.detail = detail
                }
            }
        default:
            break
        }
    }

    @MainActor
    private func applyMessageSent(_ event: ThreadEvent) {
        let fields = mergedThreadMessageFields(event)
        guard let messageIdRaw = (fields["messageId"] as? String) ?? (fields["id"] as? String),
              let roleRaw = fields["role"] as? String,
              let role = MessageRole(rawValue: roleRaw) else { return }
        let payloadText = (fields["text"] as? String) ?? ""
        let createdAt = (fields["createdAt"] as? String).flatMap(ISO8601Decoder.parse) ?? Date()
        let updatedAt = (fields["updatedAt"] as? String).flatMap(ISO8601Decoder.parse) ?? createdAt
        let streaming = (fields["streaming"] as? Bool) ?? false
        let turnId = (fields["turnId"] as? String).map { TurnID(rawValue: $0) }
        let id = MessageID(rawValue: messageIdRaw)
        let attachments = attachmentsFromMessagePayload(fields)

        // Mirror the desktop store semantics: when an existing message
        // receives a streaming event, the payload's text field is the new
        // delta to append. The completion event arrives with streaming=false
        // and text="" — keep the existing accumulated text in that case.
        if let i = messages.firstIndex(where: { $0.id == id }) {
            if streaming {
                messages[i].text += payloadText
            } else if !payloadText.isEmpty {
                messages[i].text = payloadText
            }
            messages[i].streaming = streaming
            messages[i].updatedAt = updatedAt
            if let attachments, !attachments.isEmpty {
                messages[i].attachments = attachments
            }
        } else {
            let msg = Message(id: id, role: role, text: payloadText,
                              attachments: attachments, turnId: turnId,
                              streaming: streaming, createdAt: createdAt,
                              updatedAt: updatedAt)
            messages.append(msg)
            messages.sort { $0.createdAt < $1.createdAt }
        }
    }

    /// Merges top-level event payload with nested `message` object when present.
    private func mergedThreadMessageFields(_ event: ThreadEvent) -> [String: Any] {
        var m = event.payload
        if let inner = event.payload["message"] as? [String: Any] {
            for (k, v) in inner { m[k] = v }
        }
        return m
    }

    private func attachmentsFromMessagePayload(_ fields: [String: Any]) -> [ChatImageAttachment]? {
        guard let raw = fields["attachments"] as? [[String: Any]], !raw.isEmpty else { return nil }
        let parsed = raw.compactMap { ChatImageAttachment(dictionary: $0) }
        return parsed.isEmpty ? nil : parsed
    }

    @MainActor
    private func applyMetaUpdated(_ event: ThreadEvent) {
        guard var detail = self.detail else { return }
        if let title = event.payload["title"] as? String {
            detail.title = title
        }
        if let modelDict = event.payload["modelSelection"] as? [String: Any],
           let data = try? JSONSerialization.data(withJSONObject: modelDict),
           let selection = try? JSONDecoder().decode(ModelSelection.self, from: data) {
            detail.modelSelection = selection
        }
        if let branch = event.payload["branch"] as? String {
            detail.branch = branch
        } else if event.payload["branch"] is NSNull {
            detail.branch = nil
        }
        self.detail = detail
    }

    @MainActor
    private func upsertActivity(_ activity: ThreadActivity) {
        if let i = activities.firstIndex(where: { $0.id == activity.id }) {
            activities[i] = activity
        } else {
            activities.append(activity)
        }
        if var detail = self.detail {
            if let i = detail.activities.firstIndex(where: { $0.id == activity.id }) {
                detail.activities[i] = activity
            } else {
                detail.activities.append(activity)
            }
            self.detail = detail
        }
    }

    @MainActor
    private func recomputePending() {
        pendingApprovals = PendingDerivation.pendingApprovals(from: activities)
        pendingUserInputs = PendingDerivation.pendingUserInputs(from: activities)
    }
}

extension Array where Element == Message {
    func mostRecentAssistantText() -> String? {
        last { $0.role == .assistant }?.text
    }
}

File: trifecta-android/app/src/main/java/com/belweave/trifecta/core/stores/ThreadStore.kt
package com.belweave.trifecta.core.stores

import com.belweave.trifecta.core.models.MessageID
import com.belweave.trifecta.core.models.MessageRole
import com.belweave.trifecta.core.models.Iso8601
import com.belweave.trifecta.core.models.LatestTurnState
import com.belweave.trifecta.core.models.Message
import com.belweave.trifecta.core.models.ChatImageAttachment
import com.belweave.trifecta.core.models.ModelSelection
import com.belweave.trifecta.core.models.OrchestrationSession
import com.belweave.trifecta.core.models.ProposedPlan
import com.belweave.trifecta.core.models.ProviderInteractionMode
import com.belweave.trifecta.core.models.RuntimeMode
import com.belweave.trifecta.core.models.SessionStatus
import com.belweave.trifecta.core.models.ThreadActivity
import com.belweave.trifecta.core.models.ThreadDetail
import com.belweave.trifecta.core.models.ThreadID
import com.belweave.trifecta.core.models.ThreadStreamItem
import com.belweave.trifecta.core.models.TurnID
import com.belweave.trifecta.core.models.asObjectOrNull
import com.belweave.trifecta.core.models.bool
import com.belweave.trifecta.core.models.obj
import com.belweave.trifecta.core.models.str
import com.belweave.trifecta.core.networking.StreamSubscription
import com.belweave.trifecta.core.networking.T3Client
import com.belweave.trifecta.core.networking.UploadImage
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.json.JsonObject
import java.time.Instant

class ThreadStore(val threadId: ThreadID) {

    private val _detail = MutableStateFlow<ThreadDetail?>(null)
    val detail: StateFlow<ThreadDetail?> = _detail.asStateFlow()

    private val _messages = MutableStateFlow<List<Message>>(emptyList())
    val messages: StateFlow<List<Message>> = _messages.asStateFlow()

    private val _session = MutableStateFlow<OrchestrationSession?>(null)
    val session: StateFlow<OrchestrationSession?> = _session.asStateFlow()

    private val _activities = MutableStateFlow<List<ThreadActivity>>(emptyList())
    val activities: StateFlow<List<ThreadActivity>> = _activities.asStateFlow()

    private val _proposedPlans = MutableStateFlow<List<ProposedPlan>>(emptyList())
    val proposedPlans: StateFlow<List<ProposedPlan>> = _proposedPlans.asStateFlow()

    private val _lastError = MutableStateFlow<String?>(null)
    val lastError: StateFlow<String?> = _lastError.asStateFlow()

    private val _isSending = MutableStateFlow(false)
    val isSending: StateFlow<Boolean> = _isSending.asStateFlow()

    private var subscription: StreamSubscription? = null
    private var client: T3Client? = null
    private val mutex = Mutex()

    val isTurnRunning: Boolean
        get() {
            val s = _session.value
            if (s != null && s.status == SessionStatus.RUNNING) return true
            val turn = _detail.value?.latestTurn
            if (turn != null && turn.state == LatestTurnState.RUNNING) return true
            return _isSending.value
        }

    suspend fun start(client: T3Client) {
        mutex.withLock {
            this.client = client
            subscription?.cancel()
            subscription = client.subscribeThread(threadId) { item ->
                handle(item)
            }
        }
    }

    suspend fun stop() {
        mutex.withLock {
            subscription?.cancel()
            subscription = null
            client = null
        }
    }

    suspend fun sendMessage(
        text: String,
        attachments: List<UploadImage>,
        fallbackModelSelection: ModelSelection?
    ) {
        val cli = client ?: return
        if (text.trim().isEmpty() && attachments.isEmpty()) return
        val resolved = _detail.value?.modelSelection ?: fallbackModelSelection
        if (resolved == null) {
            _lastError.value = "No model selected"
            return
        }
        _isSending.value = true
        try {
            cli.dispatchTurnStart(
                threadId = threadId,
                text = text,
                attachments = attachments,
                modelSelection = resolved,
                runtimeMode = _detail.value?.runtimeMode ?: RuntimeMode.FULL_ACCESS,
                interactionMode = _detail.value?.interactionMode ?: ProviderInteractionMode.DEFAULT
            )
        } catch (t: Throwable) {
            _lastError.value = t.message
        } finally {
            _isSending.value = false
        }
    }

    suspend fun interruptTurn() {
        val cli = client ?: return
        try {
            cli.interruptTurn(threadId, _detail.value?.latestTurn?.turnId)
        } catch (t: Throwable) {
            _lastError.value = t.message
        }
    }

    suspend fun setRuntimeMode(mode: RuntimeMode) {
        val cli = client ?: return
        if (_detail.value?.runtimeMode == mode) return
        try {
            cli.setRuntimeMode(threadId, mode)
        } catch (t: Throwable) {
            _lastError.value = t.message
        }
    }

    suspend fun setInteractionMode(mode: ProviderInteractionMode) {
        val cli = client ?: return
        if (_detail.value?.interactionMode == mode) return
        try {
            cli.setInteractionMode(threadId, mode)
        } catch (t: Throwable) {
            _lastError.value = t.message
        }
    }

    suspend fun updateModelSelection(selection: ModelSelection) {
        val cli = client ?: return
        _detail.value = _detail.value?.copy(modelSelection = selection)
        try {
            cli.updateThreadModelSelection(threadId, selection)
        } catch (t: Throwable) {
            _lastError.value = t.message
        }
    }

    fun clearLastError() {
        _lastError.value = null
    }

    private fun handle(item: ThreadStreamItem) {
        when (item) {
            is ThreadStreamItem.Snapshot -> {
                _detail.value = item.detail
                _messages.value = item.detail.messages.sortedBy { it.createdAt }
                _session.value = item.detail.session
                _proposedPlans.value = item.detail.proposedPlans
                _activities.value = item.detail.activities
            }
            is ThreadStreamItem.Event -> apply(item.event.type, item.event.payload)
        }
    }

    private fun apply(type: String, payload: JsonObject) {
        when (type) {
            "thread.message-sent" -> applyMessageSent(payload)
            "thread.session-set" -> {
                val s = OrchestrationSession.fromJson(payload.obj("session"))
                _session.value = s
                _detail.value = _detail.value?.copy(session = s)
            }
            "thread.runtime-mode-set" -> {
                val mode = RuntimeMode.fromRaw(payload.str("runtimeMode"))
                _detail.value = _detail.value?.copy(runtimeMode = mode)
            }
            "thread.interaction-mode-set" -> {
                val mode = ProviderInteractionMode.fromRaw(payload.str("interactionMode"))
                _detail.value = _detail.value?.copy(interactionMode = mode)
            }
            "thread.meta-updated" -> applyMetaUpdated(payload)
            "thread.archived" -> {
                val ts = Iso8601.parse(payload.str("archivedAt")) ?: Instant.now()
                _detail.value = _detail.value?.copy(archivedAt = ts)
            }
            "thread.unarchived" -> {
                _detail.value = _detail.value?.copy(archivedAt = null)
            }
            "thread.activity-appended" -> {
                val activity = ThreadActivity.fromJson(payload.obj("activity"))
                if (activity != null) upsertActivity(activity)
            }
            "thread.proposed-plan-upserted" -> {
                val plan = ProposedPlan.fromJson(payload.obj("proposedPlan"))
                if (plan != null) upsertPlan(plan)
            }
        }
    }

    private fun applyMessageSent(payload: JsonObject) {
        val merged = mergeMessageFields(payload)
        val rawId = merged.str("messageId") ?: merged.str("id") ?: return
        val role = MessageRole.fromRaw(merged.str("role"))
        val text = merged.str("text") ?: ""
        val createdAt = Iso8601.parse(merged.str("createdAt")) ?: Instant.now()
        val updatedAt = Iso8601.parse(merged.str("updatedAt")) ?: createdAt
        val streaming = merged.bool("streaming") ?: false
        val turnId = merged.str("turnId")?.let { TurnID(it) }
        val id = MessageID(rawId)
        val attachments = merged["attachments"]?.let { el ->
            val arr = (el as? kotlinx.serialization.json.JsonArray) ?: return@let null
            arr.mapNotNull { ChatImageAttachment.fromJson(it.asObjectOrNull()) }.takeIf { it.isNotEmpty() }
        }

        val current = _messages.value.toMutableList()
        val idx = current.indexOfFirst { it.id == id }
        if (idx >= 0) {
            val existing = current[idx]
            val newText = when {
                streaming -> existing.text + text
                text.isNotEmpty() -> text
                else -> existing.text
            }
            current[idx] = existing.copy(
                text = newText,
                streaming = streaming,
                updatedAt = updatedAt,
                attachments = if (!attachments.isNullOrEmpty()) attachments else existing.attachments
            )
        } else {
            current.add(
                Message(
                    id = id,
                    role = role,
                    text = text,
                    attachments = attachments,
                    turnId = turnId,
                    streaming = streaming,
                    createdAt = createdAt,
                    updatedAt = updatedAt
                )
            )
        }
        current.sortBy { it.createdAt }
        _messages.value = current
    }

    private fun mergeMessageFields(payload: JsonObject): JsonObject {
        val inner = payload.obj("message") ?: return payload
        val merged = HashMap<String, kotlinx.serialization.json.JsonElement>(payload)
        for ((k, v) in inner) merged[k] = v
        return JsonObject(merged)
    }

    private fun applyMetaUpdated(payload: JsonObject) {
        var d = _detail.value ?: return
        payload.str("title")?.let { d = d.copy(title = it) }
        ModelSelection.fromJson(payload.obj("modelSelection"))?.let {
            d = d.copy(modelSelection = it)
        }
        if (payload.containsKey("branch")) {
            d = d.copy(branch = payload.str("branch"))
        }
        _detail.value = d
    }

    private fun upsertActivity(activity: ThreadActivity) {
        val list = _activities.value.toMutableList()
        val idx = list.indexOfFirst { it.id == activity.id }
        if (idx >= 0) list[idx] = activity else list.add(activity)
        _activities.value = list
        _detail.value?.let { d ->
            val ds = d.activities.toMutableList()
            val di = ds.indexOfFirst { it.id == activity.id }
            if (di >= 0) ds[di] = activity else ds.add(activity)
            _detail.value = d.copy(activities = ds)
        }
    }

    private fun upsertPlan(plan: ProposedPlan) {
        val list = _proposedPlans.value.toMutableList()
        val idx = list.indexOfFirst { it.id == plan.id }
        if (idx >= 0) list[idx] = plan else list.add(plan)
        _proposedPlans.value = list
    }
}
